### PR TITLE
IA-2339  delete runtime when cryptomining activity happens

### PR DESCRIPTION
**Pull request by Qi77Qi**
This PR currently only deletes GCE VMs. We also don't delete PDs if they're using PDs, which is fine because we probably don't care so much about supporting cryptomining users well.

https://broadworkbench.atlassian.net/browse/IA-2339

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+/**
+ * This is subscriber for messages that are not published by Leonardo itself.
+ * But rather from cron jobs, or cloud logging sink (triggered from crypto-mining activity)
+ */
+class NonLeoMessageSubscriber[F[_]: Timer](gkeAlg: GKEAlgebra[F],
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala:26]: I like the idea of splitting up the `*MessageSubscriber` classes to Leo and non-Leo

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        F.unit
+      else
+        for {
+          runtimeFromGoogle <- computeService.getInstance(msg.resource.labels.projectId,
+                                                          msg.resource.labels.zone,
+                                                          InstanceName(msg.resource.labels.instanceId.toString))
+          // We mark the runtime as Deleted, and delete the instance.
+          // If instance deletion fails for some reason, it will be cleaned up by resource-validator
+          _ <- runtimeFromGoogle.traverse { instance =>
+            for {
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala:99]: Can we update a metric too? So we can make a stackdriver dashboard for this

**Comment by rtitle** [comment]
At a glance looks good, seems like it'll get the job done :)

We may want to further enhance it to send the runtime `creator` email somewhere so we can exclude these users from data warehouse/mixpanel.. but assuming we can iterate as needed.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala (hunk 2 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala:99]: there's https://github.com/DataBiosphere/leonardo/pull/1711/files/ed6b2882d121c6b0f2f4f1f25764b2c5c3058eb8#diff-1ea743458d40ca65f6c9a7f7d9178ded5967d287a3113158f4e9f713b01d930bR43.... are you thinking something more?

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala:99]: Oh missed that.. that's probably fine then

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala (hunk 1 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala:26]: yea..for leo messages, we have to define both encoder and decoder for the messages. but we shoudlnt' need to define encoder for non leo messages

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  val process: Stream[F, Unit] = subscriber.messages
+    .parEvalMapUnordered(10)(messageHandler)
+    .handleErrorWith(error => Stream.eval(logger.error(error)("Failed to initialize message processor")))
+
+  private[monitor] def messageHandler(event: Event[NonLeoMessage]): F[Unit] = {
+    val traceId = event.traceId.getOrElse(TraceId("None"))
+    for {
+      now <- nowInstant[F]
+      implicit0(ev: Ask[F, AppContext]) <- F.pure(Ask.const[F, AppContext](AppContext(traceId, now, None)))
+      _ <- metrics.incrementCounter(event.msg.messageType)
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/NonLeoMessageSubscriber.scala]: Could we name it like `s"LeoPubSub/${event.msg.messageType}"`? Just to distinguish it when making dashboards.

**Comment by Qi77Qi** [comment]
@rtitle since we have `deletedFrom` field now, we'll be able to know who did these afterwards..I'm thinking punt on BQ in this PR since I'm not sure permission wise if things will work, and this PR already has some permission thing that we need to make sure when it's rolled out to prod, so feels safer to separate the 2.

**Comment by Qi77Qi** [comment]
jenkins retest...seems a flaky unit test

**Comment by Qi77Qi** [comment]
fiab start  failed....jenkins retest

**Comment by Qi77Qi** [comment]
test failed because `leoDeadLetterTopic` doesn't exist in broard-dsde-qa...added now..jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 4]
```diff
@@ -239,6 +239,16 @@ object clusterQuery extends TableQuery(new ClusterTable(_)) {
     fullClusterQuery(baseQuery)
   }
 
+  def fullClusterRecordQueryByUniqueKey(googleProject: GoogleProject,
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: MInor: we usually name these methods `full*` when they do all the joins… maybe name this `clusterRecordQueryByUniqueKey`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala [hunk 5]
```diff
@@ -514,8 +514,13 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
       )
 
       // delete the namespace only after the helm uninstall completes
-      _ <- kubeService.deleteNamespace(dbApp.cluster.getGkeClusterId,
-                                       KubernetesNamespace(dbApp.app.appResources.namespace.name))
+      _ <- kubeService
+        .deleteNamespace(dbApp.cluster.getGkeClusterId, KubernetesNamespace(dbApp.app.appResources.namespace.name))
+        .recoverWith {
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala]: Is this recovery in wb-libs now, can it be removed here?

**Review by rtitle — APPROVED** [review]

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala (hunk 5 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala]: ah yes...forgot to clean up here

**Comment by Qi77Qi** [comment]
jenkins retest

**Pull request merged**